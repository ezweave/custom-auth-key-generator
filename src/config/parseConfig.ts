import { get, isEmpty, reduce } from 'lodash';

import { ArgTypes } from '../parseArgs';
import { camelToUpperSnake } from '../util';

export interface Config {
    headerPrefix: string,
    privateKey: string,
    publicKey: string,
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any 
const checkErrors = (errors: Array<string>, value: any, key: string): Array<string> =>
    (value === undefined || value === 'undefined') ? [
        ...errors,
        `Missing value for required field '${key}'/'${camelToUpperSnake(key)}'`, 
    ] : errors;

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/no-explicit-any 
export const parseConfig = (source: any, required = true): Config => {
    const config = {
        headerPrefix: get(source, camelToUpperSnake(ArgTypes.headerPrefix)),
        privateKey: get(source, camelToUpperSnake(ArgTypes.privateKey)),
        publicKey: get(source, camelToUpperSnake(ArgTypes.publicKey)),
    };

    const errors = reduce(
        config, checkErrors, []
    );

    if(!required || isEmpty(errors)) {
        return config;
    } else {
        throw new Error(errors.join('\n'));
    }
};

