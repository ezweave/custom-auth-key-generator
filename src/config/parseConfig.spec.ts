import { parseConfig } from './parseConfig';

describe('parseConfig', () => {
    it('parses config values from any source', () => {
        const source = {
            PRIVATE_KEY: 'Private',
            PUBLIC_KEY: 'Public',
            HEADER_PREFIX: 'Prefix',
        };
        expect(
            parseConfig(source)
        ).toMatchSnapshot();
    });
    it('throws an error if a required field is missing', () => {
        const source = {
            PRIVATE_KEY: 'Private',
        };
        expect(
         () => parseConfig(source)
        ).toThrowErrorMatchingSnapshot();
    });
    it('doesnt throw errors if required is off', () => {
        const source = {
            PRIVATE_KEY: 'Private',
        };
        expect(
         parseConfig(source, false)
        ).toMatchSnapshot();
    });
});
