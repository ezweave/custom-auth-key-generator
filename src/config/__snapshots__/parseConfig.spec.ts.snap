// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseConfig doesnt throw errors if required is off 1`] = `
Object {
  "headerPrefix": undefined,
  "privateKey": "Private",
  "publicKey": undefined,
}
`;

exports[`parseConfig parses config values from any source 1`] = `
Object {
  "headerPrefix": "Prefix",
  "privateKey": "Private",
  "publicKey": "Public",
}
`;

exports[`parseConfig throws an error if a required field is missing 1`] = `
"Missing value for required field 'headerPrefix'/'HEADER_PREFIX'
Missing value for required field 'publicKey'/'PUBLIC_KEY'"
`;
